### DISCUSSION TITLE
Automatic bank switching for easyflash

### DISCUSSION THREAD

**Discussion by adaumann** [drmortalwombat/oscar64#157] [Ideas]
Hello,
in my current project I'm using bank switching for easyflash quite intensive and it work well. If I have a function which need to be called from another bank I jump in the common memory via a proxy function and back. Example:

// Bank 1
void doSomething()
{
    proxyCallSomethingFromBank2(1);
}

// Bank 2
void callSomethingFromBank2()
{
   printf("Hello World");
}

// Main memory 0900-8000
__noinline void proxy callSomethingFromBank2(byte callBank)
{
    eflash.bank=2;
    callSomethingFromBank2();
    eflash.bank=callBank;
}

Therefore I have to write a lot of proxy functions, to allow calls between easyflash banks. I think to compiler could already provide these proxy functions (or the main area) for me. i.e. I could simply "port" existing C code without writing proxy functions. This would help me a lot. 

What do you think?

**Comment by drmortalwombat** [comment]
What about a zero overhead C++ template solution:

```
template<int back, class fn, class ... P>
__noinline auto sys_call_p(P... p)
{
	eflash.bank = __bankof(fn);
	auto r = fn(p...);
	if (back != 0xff)
		eflash.bank = back;
	return r;
}

template<class fn>
class SysCall
{
public:
	template<class ... P>
	auto operator()(P ... p) const
	{
		if (__bankof("") == __bankof(fn))
			return fn(p...);
		else
			return sys_call_p<__bankof(""), fn, P...>(p...);
	}
};

```

You can then easily promote your function to a function with proxy by:

```
SysCall<print1_>	print1;
SysCall<print2_>	print2;
SysCall<print3_>	print3;
SysCall<print4_>	print4;
SysCall<print5_>	print5;
SysCall<print6_>	print6;
```

And then invoke them from anywhere with:

```
	print1();
	print2("hello");
	print3();
	print4(5, 8);
	print5();
	print6();
```

**Reply by adaumann** [reply]
I was not aware that classes and templates are possible… will try it 
Thank you

**Comment by drmortalwombat** [comment]
I am still working on this, I will keep you posted, once it is ready to try.

**Comment by adaumann** [comment]
Impressed that C++ features are or will be possible… maybe the bank switching via templates will be added to the easyflash library. Thank you

**Comment by drmortalwombat** [comment]
New release with template support for bank switch calls is up

**Comment by adaumann** [comment]
Thank you, will try it …
